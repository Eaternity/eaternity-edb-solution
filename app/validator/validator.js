// let's' try to write the complete validator in functional style
import jsonschema from 'jsonschema'
import {curry} from 'ramda'

// Definition of product fields.

/*
These fields are mandatory and NOT allowed to pull from the linked product.
 */
const UNIQUE_FIELDS = [
  'id',
  'name',
]

/*
These fields are mandatory, but are allowed to pull from the linked product.
 */
const MANDATORY_FIELDS_FROM_LINKED_PRODUCT = [
  'nutrition-id',
  'tags',
  'perishability',
  'co2-value'
]

/*
 These fields are optional, but are allowed to pull from the linked product.
  */
export const OPTIONAL_FIELDS_FROM_LINKED_PRODUCT = [
  'fao-product-id',
  'water-scarcity-footprint-id',
  'waste-id',
  'season-begin',
  'season-end',
  'processes',
  'allergenes',
  'unit-weight',
  'density',
  'contains',
  'production-names',
  'production-values',
  'conservation-names',
  'conservation-values',
  'processing-names',
  'processing-values',
  'packaging-names',
  'packaging-values'
]

/*
Every product which can be exported shall contain this fields
 */
export const MANDATORY_FIELDS = [...UNIQUE_FIELDS, ...MANDATORY_FIELDS_FROM_LINKED_PRODUCT]

/*
All these fields are pulled from the linked product.
 */
export const ALL_FIELDS_FROM_LINKED_PRODUCT = [...MANDATORY_FIELDS_FROM_LINKED_PRODUCT, ...OPTIONAL_FIELDS_FROM_LINKED_PRODUCT]

export const orderProcesses = processes => {
  const keys = ['process', 'nutr-change-id']
  const orderedProcesses = processes.filter(process => process).map(process => {
    const orderedProcess = keys.map(key => {
      return {[key]: process[key]}
    })

    return Object.assign({}, ...orderedProcess)
  })

  return orderedProcesses
}

// TODO this is code duplication with above - how to handle that?
export const orderContains = contains => {
  const keys = ['substance', 'percentage']
  const orderedSubstances = contains.filter(substance => substance).map(substance => {
    const orderedSubstance = keys.map(key => {
      return {[key]: substance[key]}
    })

    return Object.assign({}, ...orderedSubstance)
  })

  return orderedSubstances
}

const _orderProduct = (orderProcesses, orderContains, orderedKeys, product) => {
  const orderedPairs = orderedKeys
    .filter(key => product[key] !== undefined)
    .map(key => {
      if (key === 'processes') {
        return {
          [key]: orderProcesses(product[key])
        }
      } else if (key === 'contains') {
        return {
          [key]: orderContains(product[key])
        }
      } else {
        return {[key]: product[key]}
      }
    })

  const orderedProduct = Object.assign({}, ...orderedPairs)
  return orderedProduct
}

const curriedOrderProduct = curry(_orderProduct)
export const orderProduct = curriedOrderProduct(orderProcesses, orderContains)

const _removeEmptyArrays = (orderedKeys, product) =>
  orderedKeys
    .filter(key => !(Array.isArray(product[key]) && product[key].length === 0))
    .map(key => {
      return {[key]: product[key]}
    })
    .reduce((obj, newField) => Object.assign(obj, newField), {})

export const removeEmptyArrays = curry(_removeEmptyArrays)

const _removeEmptyObjectsFromArrays = (orderedKeys, product) =>
  orderedKeys
    // I really think we should also remove empty strings...
    // .filter(key => product[key] !== '')
    // and empty objects
    // .filter(key => !(Object.keys(product[key]).length === 0 &&
    //    product[key].constructor === Object)
    // remove empty objects from arrays
    .map(key => {
      return Array.isArray(product[key])
        ? {
          [key]: product[key].filter(
              element =>
                // remove emty objects from array
                !(
                  Object.keys(element).length === 0 &&
                  element.constructor === Object
                )
            )
        }
        : {[key]: product[key]}
    })
    .filter(field => field[Object.keys(field)] !== undefined)
    .reduce((obj, newField) => Object.assign(obj, newField), {})

export const removeEmptyObjectsFromArrays = curry(_removeEmptyObjectsFromArrays)

export const addValidationSummary = product => {
  // define a default validationSummary
  const validationSummary = {
    isValid: false,
    parentProduct: '',
    brokenLinks: [],
    missingFields: [],
    missingMandatoryFields: [],
    validationErrors: []
  }
  const hasValidationSummary =
    product.hasOwnProperty('validationSummary') &&
    Object.keys(product.validationSummary).every(key => {
      return Object.keys(validationSummary).includes(key)
    })

  if (!hasValidationSummary) {
    product = {...product, validationSummary}
  }

  return product
}

const _schemaValidate = (jsonschema, addValidationSummary, schema, product) => {
  product = addValidationSummary(product)
  let {validationSummary} = product

  const validationErrors = jsonschema
    .validate(product, schema)
    .errors.map(error => {
      return error.stack
    })

  const hasValidationErrors = validationErrors.length > 0

  if (hasValidationErrors) {
    validationSummary = {...validationSummary, validationErrors}
  }

  return {...product, validationSummary}
}

const curriedSchemaValidate = curry(_schemaValidate)
export const schemaValidate = curriedSchemaValidate(
  jsonschema,
  addValidationSummary
)

const _addParentProduct = (addValidationSummary, prods, product) => {
  product = addValidationSummary(product)
  let {validationSummary} = product

  const isLinked = product.hasOwnProperty('linked-id')

  if (isLinked) {
    const parentName = prods
      .map(product => product.filename)
      .filter(filename => filename.split('-')[0] === product['linked-id'])

    validationSummary = {
      ...validationSummary,
      parentProduct: parentName[0] || ''
    }
  }

  return {...product, validationSummary}
}

const curriedAddParentProduct = curry(_addParentProduct)
export const addParentProduct = curriedAddParentProduct(addValidationSummary)

const _fillValidationSummary = (allFieldsFromLinkedProduct, MANDATORY_FIELDS, product) => {
  product = addValidationSummary(product)
  let {validationSummary} = product

  const missingFields = allFieldsFromLinkedProduct.filter(field => {
    return !product.hasOwnProperty(field)
  })

  const missingMandatoryFields = missingFields.filter(field => {
    return MANDATORY_FIELDS.includes(field)
  })

  const fieldsMissing = missingFields.length > 0
  const mandatoryFieldsMissing = missingMandatoryFields.length > 0

  if (fieldsMissing) {
    validationSummary = {...validationSummary, missingFields}
  }

  if (mandatoryFieldsMissing) {
    validationSummary = {...validationSummary, missingMandatoryFields}
  }

  return {...product, validationSummary}
}

const curriedFillValidationSummary = curry(_fillValidationSummary)
export const fillValidationSummary = curriedFillValidationSummary(
  ALL_FIELDS_FROM_LINKED_PRODUCT,
  MANDATORY_FIELDS
)

const _validateNutritionId = (addValidationSummary, nutrs, product) => {
  product = addValidationSummary(product)
  let {validationSummary} = product
  const {brokenLinks} = validationSummary

  const hasNutritionId = product.hasOwnProperty('nutrition-id')

  if (hasNutritionId) {
    const nutritionId = product['nutrition-id']
    const linkedNutritionIdExists = nutrs.some(nutrObj => {
      return nutrObj.id === nutritionId
    })

    if (!linkedNutritionIdExists) {
      validationSummary = {
        ...validationSummary,
        brokenLinks: [...brokenLinks, 'nutrition-id']
      }
    }
  }

  return {...product, validationSummary}
}

const curriedValidateNutritionId = curry(_validateNutritionId)
export const validateNutritionId = curriedValidateNutritionId(
  addValidationSummary
)

const _validateNutrChangeId = (addValidationSummary, nutrChange, product) => {
  product = addValidationSummary(product)
  let {validationSummary} = product
  const {brokenLinks} = validationSummary

  const hasProcesses = product.hasOwnProperty('processes') && product.processes

  const hasNutritionChangeId = hasProcesses
    ? product.processes.length > 0 &&
      product.processes[0].hasOwnProperty('nutr-change-id')
    : false

  if (hasNutritionChangeId) {
    const {processes} = product
    const allNutritionChangeIds = processes.map(processesObj => {
      return processesObj['nutr-change-id']
    })

    const linkedNutritionChangeIdsExist = allNutritionChangeIds.every(
      nutritionChangeId => {
        return nutrChange.some(nutrChangeObj => {
          return nutrChangeObj.id === nutritionChangeId
        })
      }
    )

    if (!linkedNutritionChangeIdsExist) {
      validationSummary = {
        ...validationSummary,
        brokenLinks: [...brokenLinks, 'nutr-change-id']
      }
    }
  }

  return {...product, validationSummary}
}

const curriedValidateNutritionChangeId = curry(_validateNutrChangeId)
export const validateNutrChangeId = curriedValidateNutritionChangeId(
  addValidationSummary
)

export const classify = product => {
  let {validationSummary} = product

  if (!validationSummary) {
    throw new Error('Cannot classify product without validationSummary')
  }

  const {
    brokenLinks,
    missingMandatoryFields,
    validationErrors
  } = validationSummary
  const hasBrokenLinks = brokenLinks.length > 0
  const hasMissingMandatoryFields = missingMandatoryFields.length > 0
  const hasValidationErrors = validationErrors.length > 0
  const isValid =
    !hasBrokenLinks && !hasMissingMandatoryFields && !hasValidationErrors
  validationSummary = {...validationSummary, isValid}
  return {...product, validationSummary}
}

const _getFieldFromParent = (prods, parentProduct, field) => {
  const validatedParentProduct = addParentProduct(prods, parentProduct)
  const {validationSummary} = validatedParentProduct
  let pulledField = {}

  const fieldExistsInParent = parentProduct.hasOwnProperty(field)
  // parentProduct is '' (falsy) when no linked-id is given or no product with
  // the linked id exists
  const isLinked = validationSummary.parentProduct

  if (fieldExistsInParent) {
    pulledField = {[field]: parentProduct[field]}
  } else if (isLinked) {
    const grandParent = prods.filter(product => {
      return product.filename === validationSummary.parentProduct
    })[0]
    const validatedGrandParent = addParentProduct(prods, grandParent)
    return _getFieldFromParent(prods, validatedGrandParent, field)
  } else {
    pulledField = {[field]: 'NOT_RESOLVABLE'}
  }

  return pulledField
}

export const getFieldFromParent = curry(_getFieldFromParent)

const _pullFieldsFromParent = (getFieldFromParent, prods, validatedProduct) => {
  const {missingFields} = validatedProduct.validationSummary

  const pulledFields = missingFields
    .map(field => getFieldFromParent(prods, validatedProduct, field))
    .filter(field => field[Object.keys(field)] !== 'NOT_RESOLVABLE')
    .reduce((prev, curr) => ({...prev, ...curr}), {})

  return pulledFields
}

const curriedPullFieldsFromParent = curry(_pullFieldsFromParent)
export const pullFieldsFromParent = curriedPullFieldsFromParent(getFieldFromParent)

const _pullAndAddFieldsFromParent = (pullFieldsFromParent, prods, product) => {
  const pulledFields = pullFieldsFromParent(prods, product)
  return {...product, ...pulledFields}
}

const curriedPullAndAddFieldsFromParent = curry(_pullAndAddFieldsFromParent)
export const pullAndAddFieldsFromParent = curriedPullAndAddFieldsFromParent(
  pullFieldsFromParent
)
