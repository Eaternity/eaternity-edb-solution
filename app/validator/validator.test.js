import path from 'path'
import jsonfile from 'jsonfile'
import {pipe} from 'ramda'
import {
  loadProduct,
  loadAllProducts,
  loadNutrs,
  loadNutrChange,
  loadProductSchema,
  removeHelperFields
} from './helpers/helpers'
import {
  orderProcesses,
  orderProduct,
  removeEmptyObjectsFromArrays,
  removeEmptyArrays,
  addValidationSummary,
  schemaValidate,
  addParentProduct,
  fillValidationSummary,
  validateNutritionId,
  validateNutrChangeId,
  classify,
  getFieldFromParent,
  pullFieldsFromParent,
  pullAndAddFieldsFromParent
} from './validator'
import {OPTIONAL_FIELDS_FROM_LINKED_PRODUCT} from './validator'
import {ALL_FIELDS_FROM_LINKED_PRODUCT} from "./validator";

// Point to fake test resources and load some standard resources
const dataDir = path.resolve(`${__dirname}`, './eaternity-edb-data-fake')
const prods = loadAllProducts(dataDir)
const nutrs = loadNutrs(dataDir)
const nutrChange = loadNutrChange(dataDir)
const productSchema = loadProductSchema(dataDir)
const orderedKeys = Object.keys(productSchema.properties)

describe('validator', () => {
  test('_orderProcesses correctly orders processes array', () => {
    const unorderedProduct = jsonfile.readFileSync(
      `${dataDir}/prods/11-unordered-prod.json`
    )
    const orderedProduct = jsonfile.readFileSync(
      `${dataDir}/prods/12-ordered-prod.json`
    )
    const unorderedProcesses = unorderedProduct.processes
    const expectedOrderedProcesses = orderedProduct.processes
    const orderedProcesses = orderProcesses(unorderedProcesses)
    expect(orderedProcesses).toEqual(expectedOrderedProcesses)
  })

  it('orderProduct orders a product according to a list of keys', () => {
    const unorderedProduct = jsonfile.readFileSync(
      `${dataDir}/prods/11-unordered-prod.json`
    )
    const expectedOrderedProduct = jsonfile.readFileSync(
      `${dataDir}/prods/12-ordered-prod.json`
    )
    const orderedProduct = orderProduct(orderedKeys)(unorderedProduct)
    expect(removeHelperFields(orderedProduct)).toEqual(
      removeHelperFields(expectedOrderedProduct)
    )
  })

  test('removeEmptyArrays removes all fields holding []', () => {
    const productWithEmptyArrays = {
      id: 1234,
      name: 'test',
      synonyms: [],
      'co2-value': 42,
      processes: []
    }

    const validateProduct = pipe(
      removeEmptyArrays(orderedKeys),
      // this is just to remove the empty fields
      orderProduct(orderedKeys)
    )

    const result = validateProduct(productWithEmptyArrays)
    expect(result).toMatchSnapshot()
    expect(Object.keys(result).length).toBe(
      Object.keys(productWithEmptyArrays).length - 2
    )
  })

  test('removeEmptyObjectsFromArrays removes [{}]', () => {
    const productWithArrayWithEmptyObject = {
      id: 15,
      processes: [{key: 'value'}, {}]
    }

    const result = removeEmptyObjectsFromArrays(orderedKeys)(
      productWithArrayWithEmptyObject
    )
    expect(result).toMatchSnapshot()
  })

  test('addValidationSummary adds validationSummary to product', () => {
    // define a validationResult
    const expectedValidationSummary = {
      isValid: false,
      parentProduct: '',
      brokenLinks: [],
      missingFields: [],
      missingMandatoryFields: [],
      validationErrors: []
    }

    const someProduct = {
      id: 1,
      name: 'Some product'
    }

    const productWithValidationSummary = addValidationSummary(someProduct)
    const {validationSummary} = productWithValidationSummary

    expect(productWithValidationSummary.id).toEqual(1)
    expect(productWithValidationSummary.name).toEqual('Some product')
    expect(validationSummary).toEqual(expectedValidationSummary)
  })

  test('addValidationSummary does not overwrite existing summary', () => {
    // define a validationResult
    const existingValidationSummary = {
      isValid: false,
      parentProduct: '2-parent-prod.json',
      brokenLinks: ['nutr-change-id'],
      missingFields: ['perishability'],
      validationErrors: []
    }
    const productWithValidationSummary = {
      id: 1,
      name: 'Some product',
      validationSummary: existingValidationSummary
    }
    const validatedProduct = addValidationSummary(productWithValidationSummary)
    const {validationSummary} = validatedProduct
    expect(validationSummary).toEqual(existingValidationSummary)
  })

  test('schemaValidate catches schema errors and adds them to summary', () => {
    const expectedValidationErrors = [
      'instance.synonyms[0] is not of a type(s) string',
      'instance.tags is not of a type(s) string',
      'instance.co2-value is not of a type(s) number'
    ]
    const productWithWrongTypes = jsonfile.readFileSync(
      `${dataDir}/prods/9-wrong-types-prod.json`
    )
    const validatedProductWithWrongTypes = schemaValidate(
      productSchema,
      productWithWrongTypes
    )
    const {validationErrors} = validatedProductWithWrongTypes.validationSummary
    expect(validationErrors).toEqual(expectedValidationErrors)
  })

  it('addParentProduct adds name of parent product to summary', () => {
    const expectedParentProduct = '2-parent-prod.json'
    const pathToChild = `${dataDir}/prods/3-child-prod.json`
    const child = jsonfile.readFileSync(pathToChild)
    const productWithParent = addParentProduct(prods, child)
    const {parentProduct} = productWithParent.validationSummary
    expect(parentProduct).toEqual(expectedParentProduct)
  })

  it('addParentProduct adds name of parent product to summary', () => {
    const expectedParentProduct = '2-parent-prod.json'
    const pathToChild = `${dataDir}/prods/3-child-prod.json`
    const child = loadProduct(pathToChild)
    const productWithParent = addParentProduct(prods, child)
    const {parentProduct} = productWithParent.validationSummary
    expect(parentProduct).toEqual(expectedParentProduct)
  })

  it('addParentProduct adds empty string as parent when not linked', () => {
    const expectedParentProduct = ''
    const pathToGrandParent = `${dataDir}/prods/1-grand-parent-prod.json`
    const grandParent = loadProduct(pathToGrandParent)
    const validatedGrandParent = addParentProduct(prods, grandParent)
    const {parentProduct} = validatedGrandParent.validationSummary
    expect(parentProduct).toEqual(expectedParentProduct)
  })

  it('addParentProduct adds empty string as parent when link broken', () => {
    const expectedParentProduct = ''
    const pathToLonelychild = `${dataDir}/prods/4-lonely-child-prod.json`
    const lonelyChild = loadProduct(pathToLonelychild)
    const validatedLonelyChild = addParentProduct(prods, lonelyChild)
    const {parentProduct} = validatedLonelyChild.validationSummary
    expect(parentProduct).toEqual(expectedParentProduct)
  })

  test('addMissingLinkedFieldsToValidationSummary adds all missing fields', () => {
    const expectedMissingFields = ALL_FIELDS_FROM_LINKED_PRODUCT
    const pathToGrandParent = `${dataDir}/prods/3-child-prod.json`
    const grandParent = loadProduct(pathToGrandParent)
    const validatedGrandParent = fillValidationSummary(grandParent)
    const {missingFields} = validatedGrandParent.validationSummary
    expect(missingFields).toEqual(expectedMissingFields)
  })

  test('addMissingLinkedFieldsToValidationSummary adds nothing when no missing fields', () => {
    const expectedMissingFields = []
    const pathToFullProduct = `${dataDir}/prods/14-full-prod.json`
    const fullProduct = loadProduct(pathToFullProduct)
    const validatedFullProduct = fillValidationSummary(fullProduct)
    const {missingFields} = validatedFullProduct.validationSummary
    expect(missingFields).toEqual(expectedMissingFields)
  })

  test('validateNutritionId finds and adds broken nutrition-id links', () => {
    const expectedBrokenLinks = ['nutrition-id']
    const productWithBrokenNutritionId = jsonfile.readFileSync(
      `${dataDir}/prods/5-broken-nutrition-id-prod.json`
    )
    const validatedProduct = validateNutritionId(
      nutrs,
      productWithBrokenNutritionId
    )

    const {brokenLinks} = validatedProduct.validationSummary
    expect(brokenLinks).toEqual(expectedBrokenLinks)
  })

  test('validateNutritionId adds nothing when nutrition-id missing', () => {
    const expectedBrokenLinks = []
    const productWithNoNutritionId = jsonfile.readFileSync(
      `${dataDir}/prods/3-child-prod.json`
    )
    const validatedProduct = validateNutritionId(
      nutrs,
      productWithNoNutritionId
    )

    const {brokenLinks} = validatedProduct.validationSummary
    expect(brokenLinks).toEqual(expectedBrokenLinks)
  })

  test('validateNutrChangeId finds and adds broken nutr-change links', () => {
    const expectedBrokenLinks = ['nutr-change-id']
    const productWithBrokenNutrChangeId = jsonfile.readFileSync(
      `${dataDir}/prods/6-broken-nutr-change-id-prod.json`
    )
    const validatedProduct = validateNutrChangeId(
      nutrChange,
      productWithBrokenNutrChangeId
    )

    const {brokenLinks} = validatedProduct.validationSummary
    expect(brokenLinks).toEqual(expectedBrokenLinks)
  })

  test('validateNutrChangeId adds nothing when nutr-change-id missing', () => {
    const expectedBrokenLinks = []
    const productWithNoNutrChangeId = jsonfile.readFileSync(
      `${dataDir}/prods/3-child-prod.json`
    )
    const validatedProduct = validateNutrChangeId(
      nutrChange,
      productWithNoNutrChangeId
    )

    const {brokenLinks} = validatedProduct.validationSummary
    expect(brokenLinks).toEqual(expectedBrokenLinks)
  })

  it('classify sets isValid to true if product is valid', () => {
    const pathToFullProduct = `${dataDir}/prods/14-full-prod.json`
    const fullProduct = loadProduct(pathToFullProduct)
    const validateProduct = pipe(
      orderProduct(orderedKeys),
      schemaValidate(productSchema),
      addParentProduct(prods),
      fillValidationSummary,
      validateNutritionId(nutrs),
      validateNutrChangeId(nutrChange),
      classify
    )
    const validatedProduct = validateProduct(fullProduct)
    expect(validatedProduct.validationSummary.isValid).toBeTruthy()
  })

  it('classify throws Error when passed product without summary', () => {
    const pathToProduct = `${dataDir}/prods/3-child-prod.json`
    const productWithoutSummary = loadProduct(pathToProduct)
    const classifyProd = () => classify(productWithoutSummary)
    expect(classifyProd).toThrow(
      'Cannot classify product without validationSummary'
    )
  })

  it('pipe yourself a validator', () => {
    const pathToFullProduct = `${dataDir}/prods/14-full-prod.json`
    const fullProduct = loadProduct(pathToFullProduct)

    const validatorPipeline = pipe(
      schemaValidate(productSchema),
      addParentProduct(prods),
      fillValidationSummary,
      validateNutritionId(nutrs),
      validateNutrChangeId(nutrChange),
      classify
    )

    const expectedValidationSummary = {
      isValid: true,
      parentProduct: '',
      brokenLinks: [],
      missingFields: [],
      missingMandatoryFields: [],
      validationErrors: []
    }

    const validatedProduct = validatorPipeline(fullProduct)
    const {validationSummary} = validatedProduct
    expect(validationSummary).toEqual(expectedValidationSummary)
  })

  it('getFieldFromParent gets field from parent', () => {
    const prods = loadAllProducts(dataDir)
    const field = 'co2-value'
    const pathToParent = `${dataDir}/prods/2-parent-prod.json`
    const parent = loadProduct(pathToParent)
    const expectedReturnValue = {'co2-value': 1}
    const getField = getFieldFromParent(prods)
    const returnValue = getField(parent, field)
    expect(returnValue).toEqual(expectedReturnValue)
  })

  it('getFieldFromParent recursively gets field from grand parent', () => {
    const prods = loadAllProducts(dataDir)
    const field = 'nutrition-id'
    const pathToParent = `${dataDir}/prods/2-parent-prod.json`
    const parent = loadProduct(pathToParent)
    const expectedReturnValue = {'nutrition-id': '1'}
    const getField = getFieldFromParent(prods)
    const returnValue = getField(parent, field)
    expect(returnValue).toEqual(expectedReturnValue)
  })

  it('pullFieldsFromParent pulls all missing fields from parent', () => {
    const prods = loadAllProducts(dataDir)
    const pathToParent = `${dataDir}/prods/2-parent-prod.json`
    const parent = loadProduct(pathToParent)
    const validateParent = pipe(addParentProduct(prods), fillValidationSummary)
    const validatedParent = validateParent(parent)
    const expectedReturnValue = {
      'nutrition-id': '1',
      'season-begin': 'from grandparent',
      'season-end': 'from grandparent'
    }
    const returnValue = pullFieldsFromParent(prods, validatedParent)
    expect(returnValue).toEqual(expectedReturnValue)
  })

  it('pullFieldsFromParent pulls fields from parent and grandparent', () => {
    const prods = loadAllProducts(dataDir)
    const pathToChild = `${dataDir}/prods/3-child-prod.json`
    const child = loadProduct(pathToChild)
    const validateChild = pipe(addParentProduct(prods), fillValidationSummary)
    const validatedChild = validateChild(child)
    const expectedReturnValue = {
      'nutrition-id': '1',
      tags: 'from, parent',
      perishability: 'from parent',
      'co2-value': 1,
      'season-begin': 'from grandparent',
      'season-end': 'from grandparent',
      processes: [{'nutr-change-id': 1, process: 'from parent'}],
      contains: [{
        substance: 'from parent',
        percentage: 4
      }
      ]
    }
    const returnValue = pullFieldsFromParent(prods, validatedChild)
    expect(returnValue).toEqual(expectedReturnValue)
  })

  it('pullAndAddFieldsFromParent pulls and adds fields from summary', () => {
    const prods = loadAllProducts(dataDir)
    const pathToChild = `${dataDir}/prods/3-child-prod.json`
    const child = loadProduct(pathToChild)
    const enhanceChild = pipe(
      addParentProduct(prods),
      fillValidationSummary,
      pullAndAddFieldsFromParent(prods),
      removeHelperFields
    )
    const expectedReturnValue = {
      id: 3,
      name: 'Child',
      'linked-id': '2',
      'nutrition-id': '1',
      tags: 'from, parent',
      perishability: 'from parent',
      'co2-value': 1,
      'season-begin': 'from grandparent',
      'season-end': 'from grandparent',
      processes: [{process: 'from parent', 'nutr-change-id': 1}],
      contains: [{
        substance: 'from parent',
        percentage: 4
      }
      ]
    }
    const returnValue = enhanceChild(child)
    expect(returnValue).toEqual(expectedReturnValue)
  })
})
