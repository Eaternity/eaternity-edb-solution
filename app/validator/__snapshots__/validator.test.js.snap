exports[`validator removeEmptyFields removes [{}] 1`] = `
Object {
  "id": 15,
}
`;
